core-backend/wasmtime/src/funcs.rs

```diff
@@ -288,27 +281,28 @@ impl<E: Ext + 'static> FuncsHandler<E> {
         let func =
             move |mut caller: Caller<'_, StoreData<E>>, payload_ptr: i32, payload_len: i32| {
                 let ext = caller.data().ext.clone();
-                ext.with(|ext: &mut E| {
+                ext.with_fallible(|ext: &mut E| {
                     let mem_wrap = get_caller_memory(&mut caller, &mem);
                     let payload = get_vec(&mem_wrap, payload_ptr as usize, payload_len as usize)?;
                     ext.reply_push(&payload).into_error_code()
                 })
-                .map_err(Trap::new)?
-                .map_err(|_| "Trapping: unable to push payload into reply")
-                .map_err(Trap::new)
+                .map_err(|_| Trap::new("Trapping: unable to push payload into reply"))
             };
         Func::wrap(store, func)
     }
 
     pub fn reply_to(store: &mut Store<StoreData<E>>, mem: WasmtimeMemory) -> Func {
         let func = move |mut caller: Caller<'_, StoreData<E>>, dest: i32| {
             let ext = &caller.data().ext;
-            let message_id = match ext.with(|ext: &mut E| ext.reply_to()).map_err(Trap::new)? {
-                Some((m_id, _)) => m_id,
-                None => return Err(Trap::new("Not running in the reply context")),
-            };
-            write_to_caller_memory(&mut caller, &mem, dest as isize as _, message_id.as_ref())
-                .map_err(Trap::new)
+            if let Some((msg_id, _)) = ext
+                .with(|ext: &mut E| ext.reply_to())
+                .map_err(|_| Trap::new("Internal error: infallible call returned Err"))?
+            {
+                write_to_caller_memory(&mut caller, &mem, dest as isize as _, msg_id.as_ref())
```

Comment:
Can we do something like
```rust
ext
    .with(|ext: &mut E| ext.reply_to())
    .map_err(|_| "Internal error: infallible call returned Err")
    .and_then(|(msg_id, _)| write_to_caller_memory(&mut caller, &mem, dest as isize as _, msg_id.as_ref()))
    .map_err(Trap:::new)?;
```

Reply:
done [762e0be](https://github.com/gear-tech/gear/pull/853/commits/762e0be06f9f42026e2db7bff5a957ed5c71e9d0)

Reply:
actually, it brought more code. Maybe should revert?
Less code could be with:
```rust
ext.with(|ext: &mut E| ext.reply_to())
                .map(|v| v.ok_or("Not running in the reply context"))
                .flatten()
                .and_then(|(msg_id, _)| {
                    write_to_caller_memory(
                        &mut caller,
                        &mem,
                        dest as isize as _,
                        msg_id.as_ref(),
                    )
                })
                .map_err(Trap::new)
```
But that map + flatten is ugly

Reply:
```rust
ext.with(|ext: &mut E| ext.reply_to())
    .transpose()
    .ok_or("Not running in the reply context")
    .flatten()
    .and_then(|(msg_id, _)| {
        write_to_caller_memory(
            &mut caller,
            &mem,
            dest as isize as _,
            msg_id.as_ref(),
        )
    })
    .map_err(Trap::new)
```
Or
```rust
ext.with(|ext: &mut E| ext.reply_to())
    .and_then(|v| v.ok_or("Not running in the reply context"))
    .and_then(|(msg_id, _)| {
        write_to_caller_memory(
            &mut caller,
            &mem,
            dest as isize as _,
            msg_id.as_ref(),
        )
    })
    .map_err(Trap::new)
```

Reply:
I like the second one